### ISSUE TITLE
Cgroup v2 freezer availability detection in RHEL 8 and derivatives

### ISSUE THREAD

**Issue by NeonSludge** [k0sproject/k0s#3069]
### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Platform

```shell
Linux 4.18.0-425.19.2.el8_7.x86_64 #1 SMP Tue Apr 4 05:30:47 EDT 2023 x86_64 GNU/Linux
NAME="AlmaLinux"
VERSION="8.7 (Stone Smilodon)"
ID="almalinux"
ID_LIKE="rhel centos fedora"
VERSION_ID="8.7"
PLATFORM_ID="platform:el8"
PRETTY_NAME="AlmaLinux 8.7 (Stone Smilodon)"
ANSI_COLOR="0;34"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:almalinux:almalinux:8::baseos"
HOME_URL="https://almalinux.org/"
DOCUMENTATION_URL="https://wiki.almalinux.org/"
BUG_REPORT_URL="https://bugs.almalinux.org/"

ALMALINUX_MANTISBT_PROJECT="AlmaLinux-8"
ALMALINUX_MANTISBT_PROJECT_VERSION="8.7"
REDHAT_SUPPORT_PRODUCT="AlmaLinux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.7"
```

### Version

`v1.27.1+k0s.0`

### Sysinfo

<details><summary>`k0s sysinfo`</summary>
<pre>
Machine ID: "c4a752dd6f4b8e4016bdec0ed3ec76d816d864de901ce912f82fa493a5d38794" (from machine) (pass)
Total memory: 3.8 GiB (pass)
Disk space available for /var/lib/k0s: 34.5 GiB (pass)
Operating system: Linux (pass)
  Linux kernel release: 4.18.0-425.19.2.el8_7.x86_64 (pass)
  Max. file descriptors per process: current: 262144 / max: 262144 (pass)
  Executable in path: modprobe: /usr/sbin/modprobe (pass)
  /proc file system: mounted (0x9fa0) (pass)
  Control Groups: version 2 (pass)
    cgroup controller "cpu": available (pass)
    cgroup controller "cpuacct": available (via cpu in version 2) (pass)
    cgroup controller "cpuset": available (pass)
    cgroup controller "memory": available (pass)
    cgroup controller "devices": available (assumed) (pass)
    cgroup controller "freezer": unavailable (rejected)
    cgroup controller "pids": available (pass)
    cgroup controller "hugetlb": available (pass)
    cgroup controller "blkio": available (via io in version 2) (pass)
  CONFIG_CGROUPS: Control Group support: built-in (pass)
    CONFIG_CGROUP_FREEZER: Freezer cgroup subsystem: built-in (pass)
    CONFIG_CGROUP_PIDS: PIDs cgroup subsystem: built-in (pass)
    CONFIG_CGROUP_DEVICE: Device controller for cgroups: built-in (pass)
    CONFIG_CPUSETS: Cpuset support: built-in (pass)
    CONFIG_CGROUP_CPUACCT: Simple CPU accounting cgroup subsystem: built-in (pass)
    CONFIG_MEMCG: Memory Resource Controller for Control Groups: built-in (pass)
    CONFIG_CGROUP_HUGETLB: HugeTLB Resource Controller for Control Groups: built-in (pass)
    CONFIG_CGROUP_SCHED: Group CPU scheduler: built-in (pass)
      CONFIG_FAIR_GROUP_SCHED: Group scheduling for SCHED_OTHER: built-in (pass)
        CONFIG_CFS_BANDWIDTH: CPU bandwidth provisioning for FAIR_GROUP_SCHED: built-in (pass)
    CONFIG_BLK_CGROUP: Block IO controller: built-in (pass)
  CONFIG_NAMESPACES: Namespaces support: built-in (pass)
    CONFIG_UTS_NS: UTS namespace: built-in (pass)
    CONFIG_IPC_NS: IPC namespace: built-in (pass)
    CONFIG_PID_NS: PID namespace: built-in (pass)
    CONFIG_NET_NS: Network namespace: built-in (pass)
  CONFIG_NET: Networking support: built-in (pass)
    CONFIG_INET: TCP/IP networking: built-in (pass)
      CONFIG_IPV6: The IPv6 protocol: built-in (pass)
    CONFIG_NETFILTER: Network packet filtering framework (Netfilter): built-in (pass)
      CONFIG_NETFILTER_ADVANCED: Advanced netfilter configuration: built-in (pass)
      CONFIG_NF_CONNTRACK: Netfilter connection tracking support: module (pass)
      CONFIG_NETFILTER_XTABLES: Netfilter Xtables support: built-in (pass)
        CONFIG_NETFILTER_XT_TARGET_REDIRECT: REDIRECT target support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_COMMENT: "comment" match support: module (pass)
        CONFIG_NETFILTER_XT_MARK: nfmark target and match support: module (pass)
        CONFIG_NETFILTER_XT_SET: set target and match support: module (pass)
        CONFIG_NETFILTER_XT_TARGET_MASQUERADE: MASQUERADE target support: unknown (warning: also tried CONFIG_IP_NF_TARGET_MASQUERADE, CONFIG_IP6_NF_TARGET_MASQUERADE)
        CONFIG_NETFILTER_XT_NAT: "SNAT and DNAT" targets support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: "addrtype" address type match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_CONNTRACK: "conntrack" connection tracking match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_MULTIPORT: "multiport" Multiple port match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_RECENT: "recent" match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_STATISTIC: "statistic" match support: module (pass)
      CONFIG_NETFILTER_NETLINK: module (pass)
      CONFIG_NF_NAT: module (pass)
      CONFIG_IP_SET: IP set support: module (pass)
        CONFIG_IP_SET_HASH_IP: hash:ip set support: module (pass)
        CONFIG_IP_SET_HASH_NET: hash:net set support: module (pass)
      CONFIG_IP_VS: IP virtual server support: module (pass)
        CONFIG_IP_VS_NFCT: Netfilter connection tracking: built-in (pass)
      CONFIG_NF_CONNTRACK_IPV4: IPv4 connetion tracking support (required for NAT): unknown (warning)
      CONFIG_NF_REJECT_IPV4: IPv4 packet rejection: module (pass)
      CONFIG_NF_NAT_IPV4: IPv4 NAT: unknown (warning)
      CONFIG_IP_NF_IPTABLES: IP tables support: module (pass)
        CONFIG_IP_NF_FILTER: Packet filtering: module (pass)
          CONFIG_IP_NF_TARGET_REJECT: REJECT target support: module (pass)
        CONFIG_IP_NF_NAT: iptables NAT support: module (pass)
        CONFIG_IP_NF_MANGLE: Packet mangling: module (pass)
      CONFIG_NF_DEFRAG_IPV4: module (pass)
      CONFIG_NF_CONNTRACK_IPV6: IPv6 connetion tracking support (required for NAT): unknown (warning)
      CONFIG_NF_NAT_IPV6: IPv6 NAT: unknown (warning)
      CONFIG_IP6_NF_IPTABLES: IP6 tables support: module (pass)
        CONFIG_IP6_NF_FILTER: Packet filtering: module (pass)
        CONFIG_IP6_NF_MANGLE: Packet mangling: module (pass)
        CONFIG_IP6_NF_NAT: ip6tables NAT support: module (pass)
      CONFIG_NF_DEFRAG_IPV6: module (pass)
    CONFIG_BRIDGE: 802.1d Ethernet Bridging: module (pass)
      CONFIG_LLC: module (pass)
      CONFIG_STP: module (pass)
  CONFIG_EXT4_FS: The Extended 4 (ext4) filesystem: module (pass)
  CONFIG_PROC_FS: /proc file system support: built-in (pass)
Error: sysinfo failed

</pre>
</details>

### What happened?

RedHat has backported quite a bit of newer features into the kernel that RHEL 8 and its derivatives like AlmaLinux and Rocky Linux are based on. This includes various cgroup v2 controllers that have been implemented in newer versions of the vanilla kernel, including the freezer and hugetlb controllers.

Unfortunately, k0s doesn't detect the cgroup freezer availability in RHEL 8 and derivatives when cgroup v2 support is enabled.
It appears that it does a simple kernel version check and expects kernels past 5.2 to have the freezer controller:
https://github.com/k0sproject/k0s/blob/c71b9fc63c9183a58d75ff40a9bdf1a420bbc3bb/internal/pkg/sysinfo/probes/linux/cgroup_v2.go#L56-L58

This obviously does not factor in RedHat's backporting work and makes it impossible to install k0s via the k0s install command or via k0sctl: they don't seem to have any way to pass the --ignore-pre-flight-checks argument to k0s controller and worker processes which would be a workaround for this issue.

### Steps to reproduce

1. Install the latest RHEL 8 release or any of its derivatives, install any available updates.
2. [Enable cgroup v2](https://access.redhat.com/solutions/6898151) (add `systemd.unified_cgroup_hierarchy=1` to the kernel command line), reboot.
3. Download the latest k0s release.
4. Confirm that cgroup v2 is being used:
```
# stat -fc %T /sys/fs/cgroup/
cgroup2fs
```
5. Confirm that the cgroup freezer controller is available and systemd can use it to freeze units (these commands must not return any errors):
```
# systemd-run --unit test sleep 1000
# systemctl freeze test.service
# systemctl thaw test.service
```
6. Run `k0s sysinfo`.
7. Confirm that the `cgroup controller "freezer": unavailable (rejected)` line is present in the command output.

### Expected behavior

k0s detects the cgroup freezer controller availability on RHEL 8-based systems.

### Actual behavior

k0s does not detect the cgroup freezer controller availability on RHEL 8-based systems.

**Comment by NeonSludge** [comment]
> Can you maybe check if the devices/freezer directories exist on your installation?

Sorry for the delayed response.

These directories do not exist in cgroup v2: it has a vastly different directory structure:

```
$ ls -la /sys/fs/cgroup/
total 0
dr-xr-xr-x  5 root root 0 May 10 09:28 .
drwxr-xr-x  7 root root 0 May 10 09:28 ..
-r--r--r--  1 root root 0 May 10 09:28 cgroup.controllers
-rw-r--r--  1 root root 0 May 10 09:28 cgroup.max.depth
-rw-r--r--  1 root root 0 May 10 09:28 cgroup.max.descendants
-rw-r--r--  1 root root 0 May 10 09:28 cgroup.procs
-r--r--r--  1 root root 0 May 10 09:28 cgroup.stat
-rw-r--r--  1 root root 0 May 10 09:30 cgroup.subtree_control
-rw-r--r--  1 root root 0 May 10 09:28 cgroup.threads
-rw-r--r--  1 root root 0 May 10 09:28 cpu.pressure
-r--r--r--  1 root root 0 May 10 09:28 cpuset.cpus.effective
-r--r--r--  1 root root 0 May 10 09:28 cpuset.mems.effective
-r--r--r--  1 root root 0 May 10 09:28 cpu.stat
drwxr-xr-x  2 root root 0 May 10 09:28 init.scope
-rw-r--r--  1 root root 0 May 10 09:28 io.pressure
-r--r--r--  1 root root 0 May 10 09:28 io.stat
-r--r--r--  1 root root 0 May 10 09:28 memory.numa_stat
-rw-r--r--  1 root root 0 May 10 09:28 memory.pressure
-r--r--r--  1 root root 0 May 10 09:28 memory.stat
drwxr-xr-x 69 root root 0 May 10 09:28 system.slice
drwxr-xr-x  3 root root 0 May 10 09:48 user.slice
```

I can think of two ways to check for cgroup v2 freezer availability:
1. Check via systemd commands. This is probably too clunky.
2. Look for freezer v2 control files in some cgroup. Not sure how reliable this would be.
```
stat -c '%n: %F' /sys/fs/cgroup/system.slice/cgroup.freeze
/sys/fs/cgroup/system.slice/cgroup.freeze: regular empty file
```

**Comment by twz123** [comment]
The current draft PR contains the second variant. It tries to check for the cgroup.freeze file in the cgroup of k0s itself. IIUC the only reason for this to not work correctly is when k0s is part of the root cgroup, which doesn't have that control file.

The device controller doesn't have any control files, it has a BPF-only interface. From my current understanding, the only way to detect that feature would be to actually try to attach some dummy device filter BPF program to a cgroup. I've experimented a bit. It's actually doable and not as unwieldy as it sounds (thanks to containerd's cgroup go library). The main caveat is that this check requires elevated privileges. This is probably not a problem per se for the pre-flight checks, as k0s spawns containerd, which in turn needs those privileges as well. But this is a limitation for the `sysinfo` subcommand, as this is not necessarily run with elevated privileges, and maybe for some more exotic setups with external CRIs.

Which leads me to the following proposal:
* The freezer check is more straight forward. If k0s runs in the root cgroup (which won't be the case when it's run via systemd or OpenRC), issue a warning rather than rejecting the probe.
* For the devices controller, try to create a dummy cgroup and attach a device filter BPF program to it. Handle EPERM errors in a specific way: If any of the steps in the check fail with EPERM, reject the probe if k0s is run as root, or issue a warning otherwise.

**Comment by NeonSludge** [comment]
> Which leads me to the following proposal:
> 
> The freezer check is more straight forward. If k0s runs in the root cgroup (which won't be the case when it's run via systemd or OpenRC), issue a warning rather than rejecting the probe.
> For the devices controller, try to create a dummy cgroup and attach a device filter BPF program to it. Handle EPERM errors in a specific way: If any of the steps in the check fail with EPERM, reject the probe if k0s is run as root, or issue a warning otherwise.

That sounds reasonable.
I've tested the changes from your PR on an AlmaLinux 8.7 machine with cgroup v2 enabled and the freezer check now passes.

**Comment by twz123** [comment]
Will be released as part of k0s 1.29.
